Framework/LiveData/src/Kafka/KafkaHistoStreamDecoder.cpp

```diff
@@ -0,0 +1,238 @@
+// Mantid Repository : https://github.com/mantidproject/mantid
+//
+// Copyright &copy; 2008 ISIS Rutherford Appleton Laboratory UKRI,
+//     NScD Oak Ridge National Laboratory, European Spallation Source
+//     & Institut Laue - Langevin
+// SPDX - License - Identifier: GPL - 3.0 +
+#include "MantidLiveData/Kafka/KafkaHistoStreamDecoder.h"
+#include "MantidAPI/AlgorithmManager.h"
+#include "MantidAPI/Axis.h"
+#include "MantidAPI/Run.h"
+#include "MantidAPI/WorkspaceFactory.h"
+#include "MantidDataObjects/WorkspaceCreation.h"
+#include "MantidGeometry/Instrument/DetectorInfo.h"
+#include "MantidHistogramData/BinEdges.h"
+#include "MantidKernel/Logger.h"
+#include "MantidKernel/OptionalBool.h"
+#include "MantidKernel/WarningSuppressions.h"
+#include "MantidLiveData/Exception.h"
+
+GNU_DIAG_OFF("conversion")
+#include "private/Schema/hs00_event_histogram_generated.h"
+GNU_DIAG_ON("conversion")
+
+namespace {
+/// Logger
+Mantid::Kernel::Logger g_log("KafkaHistoStreamDecoder");
+} // namespace
+
+namespace Mantid {
+namespace LiveData {
+
+// -----------------------------------------------------------------------------
+// Public members
+// -----------------------------------------------------------------------------
+
+/**
+ * Constructor
+ * @param broker A reference to a Broker object for creating topic streams
+ * @param histoTopic The name of the topic streaming the histo data
+ * @param spDetTopic The name of the topic streaming the spectrum-detector
+ * run mapping
+ */
+KafkaHistoStreamDecoder::KafkaHistoStreamDecoder(
+    std::shared_ptr<IKafkaBroker> broker, const std::string &histoTopic,
+    const std::string &instrumentName)
+    : m_broker(broker), m_histoTopic(histoTopic),
+      m_instrumentName(instrumentName), m_histoStream(), m_workspace(),
```

Comment:
Is there any check on the histoTopic or instrumentName being valid? Do we need to check that `KafkaHistoStreamDecoder` is not being initialised with junk?

Reply:
I think what would happen with an invalid instrument name is `createBufferWorkspace` would throw an exception due to failing to load the instrument, which would then be caught in `KafkaHistoListener::connect`.

Not sure how we would detect an invalid histoTopic... what makes it invalid? If it doesn't exist yet, I think Kafka will just create one when we connect to it. We just won't get any data unless a producer connects to it as well.

Reply:
We've disabled "automatic" topic creation for the production cluster, so `rdkafka` would return an error code when you try to `assign` the non-existent topic.

Reply:
Is there a way to query whether a topic exists?

Reply:
Looks like it is already handled by the subscriber class: https://github.com/mantidproject/mantid/blob/1dc90bf3242ef7201883e96adf85eca2323ffb72/Framework/LiveData/src/Kafka/KafkaTopicSubscriber.cpp#L310